packages/warehouses/src/warehouseClients/SnowflakeWarehouseClient.ts

```diff
@@ -226,33 +226,77 @@ export class SnowflakeWarehouseClient implements WarehouseClient {
         await this.runQuery('SELECT 1');
     }
 
+    private async runTableCatalogQuery(
+        database: string,
+        schema: string,
+        table: string,
+    ) {
+        let connection: Connection;
+        const sqlText = `SHOW COLUMNS IN TABLE ${table}`;
+        try {
+            connection = createConnection({
+                ...this.connectionOptions,
+                schema,
+                database,
+            });
+            await Util.promisify(connection.connect)();
+        } catch (e) {
+            throw new WarehouseConnectionError(`Snowflake error: ${e.message}`);
+        }
+        try {
+            return await this.executeStatement(connection, sqlText);
+        } catch (e) {
+            // Ignore error and let UI show invalid table
+            return undefined;
+        } finally {
+            // todo: does this need to be promisified? uncaught error in callback?
+            connection.destroy((err) => {
+                if (err) {
+                    throw new WarehouseConnectionError(err.message);
+                }
+            });
+        }
+    }
+
     async getCatalog(
         config: {
             database: string;
             schema: string;
             table: string;
         }[],
     ) {
-        const sqlText = 'SHOW COLUMNS IN ACCOUNT';
-        const { rows } = await this.runQuery(sqlText);
-        return rows.reduce<WarehouseCatalog>((acc, row) => {
-            const match = config.find(
-                ({ database, schema, table }) =>
-                    database.toLowerCase() ===
-                        row.database_name.toLowerCase() &&
-                    schema.toLowerCase() === row.schema_name.toLowerCase() &&
-                    table.toLowerCase() === row.table_name.toLowerCase(),
-            );
-            // Unquoted identifiers will always be
-            if (row.kind === 'COLUMN' && !!match) {
-                acc[match.database] = acc[match.database] || {};
-                acc[match.database][match.schema] =
-                    acc[match.database][match.schema] || {};
-                acc[match.database][match.schema][match.table] =
-                    acc[match.database][match.schema][match.table] || {};
-                acc[match.database][match.schema][match.table][
-                    row.column_name
-                ] = mapFieldType(JSON.parse(row.data_type).type);
+        const tablesMetadataPromises = config.map(
+            ({ database, schema, table }) =>
+                this.runTableCatalogQuery(database, schema, table),
+        );
+
+        const tablesMetadata = await Promise.all(tablesMetadataPromises);
```

Comment:
Did you performance test this? Since we're going from a single query to many queries, I feel that this could be much slower.

Reply:
![Screenshot 2023-01-26 at 09 51 25](https://user-images.githubusercontent.com/9117144/214806670-5c250ec3-c6ab-40aa-8b95-06e05b0ca926.png)

The previous query takes between 1.2 to 3 seconds while the multiple queries take 50ms each, in total 350 seconds. 

Is hard to calculate a value for the performance improvement since this depends highly on the user warehouse. We were requesting everything that wasn't related to the dbt project.

Reply:
beautiful